### ISSUE TITLE
Create possibility to perform an internal forward to the consent page

### ISSUE THREAD

**Issue by Chr3is** [spring-projects/spring-authorization-server#1232]
**Expected Behavior**
If the AuthenticationProvider decided, that a user has to give consent, the user is internally forwarded to the consent page without an additional redirect. The URI does not change in the browser.

**Current Behavior**
If consent is required, the user is redirected to the provided consent url. The browser performs an additional request.

**Context**
We would like to get rid of this additional redirect and perform an internal forward to the consent controller with the original authorization request.

Possible approach:
Make the redirectStrategy in the OAuth2AuthorizationEndpointFilter configurable to be able to perform a forward if the provided url is the consent url in a custom implementation.

**Comment by jgrandja** [comment]
@Chr3is 

> We would like to get rid of this additional redirect and perform an internal forward to the consent controller with the original authorization request.

Can you please provide details on why you require a forward? Why is the redirect insufficient for your use case?

**Comment by Chr3is** [comment]
@jgrandja 
A user performs the OIDC auth code flow. After the login, the user's original auth request is replayed as authenticated user. The authentication provider decides, that the user has to give explicit scope/claim approval. If the approval is done within the same system, the approval page can be displayed directly in the browser (with an internal forward). In our case this flow is session based and we don't need the clientId, state & scopes as redirect parameters directly. **We want to minimize the count of redirects as much as possible.** In https://github.com/spring-attic/spring-security-oauth/blob/main/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/endpoint/AuthorizationEndpoint.java#L344 there was no need to perform an explicit redirect to the consent page.

**Comment by jgrandja** [comment]
@Chr3is 

> We want to minimize the count of redirects as much as possible.

I'm trying to understand why you want to minimize the redirects? Is this causing an issue? Or is it simply a preference?

This is the first time we received this enhancement request so it is unlikely we would provide this enhancement at this point since it wouldn't be widely used.

As an FYI, we try to minimize exposing public API's until there is a greater need for it from our users. Exposing public APIs restricts our flexibility to change things freely and we need to support it over a longer term based on our support cycle. So we only expose APIs if and when there a greater demand from our users.

At this point, it is unlikely we would expose `redirectStrategy`. Furthermore, an **Authorization Request** and **Authorization Consent Request** are **distinct request flows** so maintaining the Authorization Request parameters (via a forward) at the Authorization Consent screen doesn't make sense. For example, the `redirect_uri` and `code_challenge` parameters are not relevant in the context of the Authorization Consent screen.

**Comment by Chr3is** [comment]
@jgrandja It's a preference and we would like to keep the old behavior, which was implemented with the deprecated spring security oauth2 AuthorizationEndpoint, which performed an internal forward. A forward is more performant than a redirect which requires the browser to perform an additional request. At the moment the consent endpoint retrieves the clientId, state and scope. With the state we can perform the lookup for the stored authorization which was created by the AuthRequestAuthprovider. However if custom handling like claims where implemented, they are missing in the consent redirect and this is in my opinion not coherent. It should be at least possible for the developer how the consent handling is done. And this includes how the consent (if required) is handled eg. a redirect (with additional parameters) or an internal forward is performed.

**Comment by jgrandja** [comment]
@Chr3is 

> It's a preference and we would like to keep the old behavior, which was implemented with the deprecated spring security oauth2 AuthorizationEndpoint, which performed an internal forward

I can appreciate that you would prefer to keep with the existing logic but as you are aware Spring Authorization Server is a complete re-write from the legacy Authorization Server so changes are inevitable when migrating to Spring Authorization Server.

> With the state we can perform the lookup for the stored authorization which was created by the AuthRequestAuthprovider. However if custom handling like claims where implemented, they are missing in the consent redirect and this is in my opinion not coherent.

All the information related to the "in-flight" authorization is available to the consent controller after the redirect. Please see `AuthorizationConsentController`:

https://github.com/spring-projects/spring-authorization-server/blob/9c1ec348afcaa5c68d058d0b9b38fda405393edc/samples/demo-authorizationserver/src/main/java/sample/web/AuthorizationConsentController.java#L41

It demonstrates retrieval of `OAuth2AuthorizationConsent` for customizing the consent screen.

For your use case, you would inject `OAuth2AuthorizationService` and lookup the `OAuth2Authorization` via `authorizationService.findByToken(state, new OAuth2TokenType(OAuth2ParameterNames.STATE))`. Everything you need will be available in `OAuth2Authorization` to perform custom processing.

I'll leave this open for now and we'll see if other user's are looking for this capability.

**Comment by Chr3is** [comment]
@jgrandja Yes, there's no problem in implementing the consent controller. However, for me, it feels a little bit inconsistent that the scope is passed as redirect parameter but other additional params (claims for example) have to be taken from the in-flight authorization. Whats the purpose of scope redirect param at all? I like to optimize things where possible and at this point, a forward could do that. But thanks for the discussion! I appreciate that. Lets keep this open.

**Comment by amrnyang** [comment]
@jgrandja Hello，joe. I encountered a problem when I was studying the spring authorization server samples. When I used demo-authorizationserver, message-resource, and demo-client, I found that when I logged in with user1, there was personal information authorization( consent page) and client authorization(consent page), and they were continuous. When I logged in with user2, only personal information authorization appeared once, and client information authorization did not appear. I did not change any code, but just added a user2. Why does this happen? Sorry, I don't know if this is worth submitting an issue, so I can only leave a message under the question about consent page. Thank you.

**Comment by VictorMolina** [comment]
Hi @jgrandja! I wanted to override all the handlers to build a REST API for authentication. However, the redirect from the consent step is an issue because it makes the client follow redirects instead of receiving a direct JSON response. It would be great if, in the future, Spring Authorization Server could be used as a REST API for SPAs.
